pkg/addonmanager/controllers/agentdeploy/hookcontroller.go

```diff
@@ -0,0 +1,301 @@
+		c.cache.removeCache(preDeleteHookWorkName(addonName), clusterName)
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if !managedCluster.DeletionTimestamp.IsZero() {
+		// managed cluster is deleting, remove cache.
+		c.cache.removeCache(preDeleteHookWorkName(addonName), clusterName)
+		return nil
+	}
+
+	managedClusterAddon, err := c.managedClusterAddonLister.ManagedClusterAddOns(clusterName).Get(addonName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			c.cache.removeCache(preDeleteHookWorkName(addonName), clusterName)
+			return nil
+		}
+		return err
+	}
+	managedClusterAddonCopy := managedClusterAddon.DeepCopy()
+	owner := metav1.NewControllerRef(managedClusterAddon, addonapiv1alpha1.GroupVersion.WithKind("ManagedClusterAddOn"))
+
+	objects, err := agentAddon.Manifests(managedCluster, managedClusterAddon)
+	if err != nil {
+		return err
+	}
+	if len(objects) == 0 {
+		return nil
+	}
+	_, hookWork, err := buildManifestWorkFromObject(clusterName, addonName, objects)
+	if err != nil {
+		return err
+	}
+
+	if hookWork == nil {
+		if !hasFinalizer(managedClusterAddonCopy.GetFinalizers(), constants.PreDeleteHookFinalizer) {
+			return nil
+		}
+		finalizer := removeFinalizer(managedClusterAddonCopy.Finalizers, constants.PreDeleteHookFinalizer)
+		managedClusterAddonCopy.SetFinalizers(finalizer)
+		_, err = c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).Update(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+		return err
+	}
+
+	// has hook manifestWork
+	if !hasFinalizer(managedClusterAddonCopy.GetFinalizers(), constants.PreDeleteHookFinalizer) {
+		managedClusterAddonCopy.Finalizers = append(managedClusterAddonCopy.Finalizers, constants.PreDeleteHookFinalizer)
+		_, err = c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).Update(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+		return err
+	}
+
+	if managedClusterAddon.DeletionTimestamp.IsZero() {
+		return nil
+	}
+
+	// apply hookWork when addon is deleting
+	hookWork.OwnerReferences = []metav1.OwnerReference{*owner}
+	hookWork, err = c.applyWork(ctx, hookWork)
+	if err != nil {
+		meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+			Type:    "HookManifestCompleted",
+			Status:  metav1.ConditionFalse,
+			Reason:  "ManifestWorkApplyFailed",
+			Message: fmt.Sprintf("failed to apply manifestwork %v: %v", hookWork.Name, err),
+		})
+		c.addonClient.AddonV1alpha1().ManagedClusterAddOns(managedClusterAddonCopy.Namespace).UpdateStatus(
+			ctx, managedClusterAddonCopy, metav1.UpdateOptions{})
+		return err
+	}
+
+	meta.SetStatusCondition(&managedClusterAddonCopy.Status.Conditions, metav1.Condition{
+		Type:    "HookManifestCompleted",
+		Status:  metav1.ConditionFalse,
+		Reason:  "HookManifestIsNotCompleted",
+		Message: fmt.Sprintf("hook manifestWork %v is not completed.", hookWork.Name),
+	})
+
+	completed := hookWorkIsCompleted(hookWork)
```

Comment:
I think you should check completed and remove finalizer  and return at first. It will remove the addon and the following status update will not be needed.

Then you can check whether there is an apply error or completed with switch.

Reply:
something like

```go
hookWork, applyErr = c.applyWork(ctx, hookWork)
completed := hookWorkIsCompleted(hookWork)

if completed && hasFinalizer {
  return removeFinalizer()
}

switch {
case applyErr != nil:
case !completed
}

//deepEqual and update status
```

Reply:
done. this is better. thanks.